gax/src/main/java/com/google/api/gax/tracing/TracedBatchedContextCallable.java

```diff
@@ -0,0 +1,118 @@
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google LLC nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.tracing;
+
+import com.google.api.core.ApiFuture;
+import com.google.api.core.ApiFutures;
+import com.google.api.core.InternalApi;
+import com.google.api.gax.batching.BatchedCallContext;
+import com.google.api.gax.rpc.ApiCallContext;
+import com.google.api.gax.rpc.UnaryCallable;
+import com.google.api.gax.tracing.ApiTracerFactory.OperationType;
+import com.google.common.base.Preconditions;
+import com.google.common.util.concurrent.MoreExecutors;
+
+/**
+ * This callable wraps a batching callable chain in an {@link ApiTracer} and annotates {@link
+ * BatchedCallContext} batching context data.
+ *
+ * <p>For internal use only.
+ */
+@InternalApi("For internal use by google-cloud-java clients only")
+public class TracedBatchedContextCallable<RequestT, ResponseT>
+    extends UnaryCallable<RequestT, ResponseT> {
+
+  private final ApiTracerFactory tracerFactory;
+  private ApiCallContext baseCallContext;
+  private final SpanName spanName;
+  private final UnaryCallable<RequestT, ResponseT> innerCallable;
+
+  public TracedBatchedContextCallable(
+      UnaryCallable<RequestT, ResponseT> innerCallable,
+      ApiCallContext callContext,
+      ApiTracerFactory tracerFactory,
+      SpanName spanName) {
+    this.baseCallContext = Preconditions.checkNotNull(callContext);
+    this.tracerFactory = Preconditions.checkNotNull(tracerFactory);
+    this.spanName = Preconditions.checkNotNull(spanName);
+    this.innerCallable = Preconditions.checkNotNull(innerCallable);
+  }
+
+  /**
+   * Creates an {@link ApiTracer} and annotates batching context data. Performs a call
+   * asynchronously.
+   */
+  public ApiFuture<ResponseT> futureCall(RequestT request, BatchedCallContext batchedCallContext) {
```

Comment:
Can batchedCallContext just become a part of ApiCallContext? So we will not have to have two methods here, one of which (this one) does not even fit in the overall callables hierarchy (the idea is that callables have their implementation of futureCall(request, apiCallContext), but don't have their own special contexts.

Reply:
Moving this to ApiCallContext is an option. There are 2 approaches:
1. allowing ApiCallContext to generically store random data (similar to CallOptions). This will have a bit of overhead and makes this a lot more dynamically typed and I think harder to debug
2. adding a specific getter/setter for batching context - this will make a generic construct like ApiContext have to know about specific things like batching that not relevant most of the time.

Localizing the interaction between a batcher and the next callable seems to be the least of the 3 evils